File: app/schemas/order.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from app.models import OrderStatus


class OrderItemCreate(BaseModel):
    product_id: uuid.UUID
    quantity: int = Field(gt=0)
    sn_codes: list[str] = Field(default_factory=list)


class OrderCreate(BaseModel):
    customer_id: uuid.UUID
    store_id: uuid.UUID
    items: list[OrderItemCreate] = Field(min_length=1)

    @model_validator(mode="after")
    def validate_items(self) -> "OrderCreate":
        if not self.items:
            raise ValueError("At least one order item is required")
        return self


class OrderItemRead(BaseModel):
    id: uuid.UUID
    product_id: uuid.UUID
    product_name: str
    sku: str
    quantity: int
    unit_price: Decimal
    line_total: Decimal
    sn_codes: list[str] = Field(default_factory=list)


class OrderRead(BaseModel):
    id: uuid.UUID
    customer_id: uuid.UUID
    store_id: uuid.UUID
    total_amount: Decimal
    status: OrderStatus
    created_at: datetime
    items: list[OrderItemRead]


class OrderListItem(BaseModel):
    class Item(BaseModel):
        class SerialDetail(BaseModel):
            sn_code: str
            warranty_ends_at: datetime | None = None

        product_name: str
        sku: str
        quantity: int
        unit_price: Decimal
        serial_details: list[SerialDetail] = Field(default_factory=list)

    id: uuid.UUID
    customer_id: uuid.UUID
    customer_name: str
    store_id: uuid.UUID
    store_name: str
    total_amount: Decimal
    status: OrderStatus
    created_at: datetime
    items: list[Item]

File: frontend/src/services/order.ts
import { request } from '../utils/request';

export interface CreateOrderItemPayload {
  product_id: string;
  quantity: number;
  sn_codes: string[];
}

export interface CreateOrderPayload {
  customer_id: string;
  store_id: string;
  items: CreateOrderItemPayload[];
}

export interface IOrderItem {
  id: string;
  product_id: string;
  product_name: string;
  sku: string;
  quantity: number;
  unit_price: number | string;
  line_total: number | string;
  sn_codes: string[];
}

export interface IOrder {
  id: string;
  customer_id: string;
  store_id: string;
  total_amount: number | string;
  status: 'paid' | 'returned' | 'cancelled';
  created_at: string;
  items: IOrderItem[];
}

export interface IOrderListItem {
  items: Array<{
    product_name: string;
    sku: string;
    quantity: number;
    unit_price: number | string;
    serial_details: Array<{
      sn_code: string;
      warranty_ends_at: string | null;
    }>;
  }>;
  id: string;
  customer_id: string;
  customer_name: string;
  store_id: string;
  store_name: string;
  total_amount: number | string;
  status: 'paid' | 'returned' | 'cancelled';
  created_at: string;
}

export interface GetOrdersParams {
  customerName?: string;
  orderId?: string;
}

export async function createOrder(payload: CreateOrderPayload): Promise<IOrder> {
  const response = await request.post<IOrder>('/api/orders', payload);
  return response.data;
}

export async function getOrders(params: GetOrdersParams = {}): Promise<IOrderListItem[]> {
  const response = await request.get<IOrderListItem[]>('/api/orders', {
    params: {
      customer_name: params.customerName?.trim() || undefined,
      order_id_prefix: params.orderId?.trim() || undefined,
    },
  });
  return response.data;
}

export async function returnOrder(orderId: string): Promise<IOrder> {
  const response = await request.post<IOrder>(`/api/orders/${orderId}/return`);
  return response.data;
}
